Lib/tempfile.py

```diff
@@ -822,8 +822,17 @@ def resetperms(path):
 
                     try:
                         _os.unlink(path)
+                    except IsADirectoryError:
+                        cls._rmtree(path, ignore_errors=ignore_errors)
                     # PermissionError is raised on FreeBSD for directories
-                    except (IsADirectoryError, PermissionError):
+                    except PermissionError:
+                        # bpo-43153: Calling _rmtree again may
+                        # raise NotADirectoryError and mask the PermissionError.
+                        # So we must re-raise the current PermissionError.
+                        if not _os.path.isdir(path):
```

Comment:
`os.path.isdir()` returns True for a symlink to directory. Either check `os.path.islink()` before `os.path.isdir()` or use `os.stat()` explicitly.

Reply:
Failing to delete a symlink should be rare in a temp directory. But if it's worth handling here, then it should be consistent with `shutil._rmtree_isdir()`. Unfortunately the latter is written for `os.DirEntry` instances. Here's an updated implementation for `shutil` that can also check a path. I factored out the check for a mount point into a common `_st_ismount(st)` function, which avoids duplicated code and, I think, makes it easier to read.

```python
if not hasattr(os.stat_result, 'st_reparse_tag'):
    def _rmtree_islink(path):
        return os.path.islink(path)

    def _rmtree_isdir(path):
        try:
            if isinstance(path, os.DirEntry):
                return entry.is_dir(follow_symlinks=False)
            return stat.S_ISDIR(os.lstat(path).st_mode)
        except OSError:
            return False
else:
    # In general mount points in Windows are directories, but os.unlink()
    # supports them, so rmtree() handles them as symlinks.
    def _st_ismount(st):
        return (st.st_file_attributes & stat.FILE_ATTRIBUTE_REPARSE_POINT
                and st.st_reparse_tag == stat.IO_REPARSE_TAG_MOUNT_POINT)

    def _rmtree_islink(path):
        try:
            if isinstance(path, os.DirEntry):
                st = entry.stat(follow_symlinks=False)
            else:
                st = os.lstat(path)
            return stat.S_ISLNK(st.st_mode) or _st_ismount(st)
        except OSError:
            return False

    def _rmtree_isdir(path):
        try:
            if isinstance(path, os.DirEntry):
                st = entry.stat(follow_symlinks=False)
            else:
                st = os.lstat(path)
            return stat.S_ISDIR(st.st_mode) and not _st_ismount(st)
        except OSError:
            return False
```